src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.ReadAhead.cs

```diff
@@ -42,20 +41,14 @@ internal static bool DoSingleValueReadWithReadAhead(ref Utf8JsonReader reader, r
 
             // Perform the actual read-ahead.
             JsonTokenType tokenType = reader.TokenType;
-            if (tokenType == JsonTokenType.StartObject || tokenType == JsonTokenType.StartArray)
+            if (tokenType is JsonTokenType.StartObject or JsonTokenType.StartArray)
             {
                 // Attempt to skip to make sure we have all the data we need.
                 bool complete = reader.TrySkip();
 
                 // We need to restore the state in all cases as we need to be positioned back before
                 // the current token to either attempt to skip again or to actually read the value.
-
-                reader = new Utf8JsonReader(reader.OriginalSpan.Slice(checked((int)initialReaderBytesConsumed)),
```

Comment:
Similarly, this change checkpoints the reader without assuming that the instance is reading from a ReadOnlySpan source.